### PR TITLE
[AutoParallel] Add op_role_guard for PIR

### PR DESCRIPTION
**Pull request by JZ-LIANG**
### PR Category

Auto Parallel

### PR Types

Improvements

### Description

Add op_role_guard for PIR to Optimize Graph compilation for Pipeline logic 
Reduce the DAG Rewrite Compilation time in pir parallel

compile time | num_layer=32 | num_layer=64
-- | -- | --
before PR | 109 | 432
After PR | 10 | 15

Pcard-67164

### CODE REVIEW

File: paddle/pir/include/core/builder.h [hunk 1]
```diff
@@ -172,6 +174,10 @@ class Builder {
   InsertionPoint insertion_point_;
 
   bool forbid_insert_without_position_;
+
+  // by now the op_role is used by autoparallel for distinguish the op in fw,
+  // bw, opt region.
+  int current_op_role_ = -1;
```
Comment (winter-wang) [inline paddle/pir/include/core/builder.h]:
```suggestion
  int op_role_ = -1;
```

File: paddle/pir/include/core/builder.h [hunk 2]
```diff
@@ -115,7 +115,9 @@ class Builder {
         common::errors::PreconditionNotMet("argument of block is nullptr"));
     set_insertion_point(block, block->end());
   }
-
+  /// Set/Get the op_role
+  void SetCurrentOpRole(int op_role) { current_op_role_ = op_role; }
```
Comment (winter-wang) [inline paddle/pir/include/core/builder.h]:
```suggestion
  void set_op_role(int op_role) { op_role_ = op_role; }
```

File: paddle/pir/include/core/builder.h [hunk 3]
```diff
@@ -115,7 +115,9 @@ class Builder {
...
+  int GetCurrentOpRole() const { return current_op_role_; }
```
Comment (winter-wang) [inline paddle/pir/include/core/builder.h]:
```suggestion
  int op_role() const { return op_role_; }
```

Comment (JZ-LIANG) [reply paddle/pir/include/core/builder.h]: fixed

File: paddle/pir/include/core/builder.h (hunk 2 above)
Comment (JZ-LIANG) [reply paddle/pir/include/core/builder.h]: fixed

File: paddle/pir/include/core/builder.h (hunk 1 above)
Comment (JZ-LIANG) [reply paddle/pir/include/core/builder.h]: fixed

**Review by winter-wang — DISMISSED** [review]

**Review by winter-wang — APPROVED** [review]

**Pull request merged**